src/Common/src/TypeSystem/IL/Stubs/StructMarshallingThunk.cs

```diff
@@ -0,0 +1,253 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using Internal.TypeSystem;
+using Internal.TypeSystem.Interop;
+using System.Collections.Generic;
+using Debug = System.Diagnostics.Debug;
+
+namespace Internal.IL.Stubs
+{
+    internal enum StructMarshallingThunkType : byte
+    {
+        MangedToNative = 1,
+        NativeToManage = 2,
+        Cleanup = 4
+    };
+
+    internal class StructMarshallingThunk : ILStubMethod
+    {
+        internal readonly TypeDesc ManagedType;
+        internal readonly NativeStructType NativeType;
+        internal readonly StructMarshallingThunkType ThunkType;
+        private  InteropStateManager _interopStateManager;
+        private TypeDesc _owningType;
+
+        public StructMarshallingThunk(TypeDesc owningType, TypeDesc managedType, StructMarshallingThunkType thunkType, InteropStateManager interopStateManager)
+        {
+            _owningType = owningType;
+            ManagedType = managedType;
+            _interopStateManager = interopStateManager;
+            NativeType = _interopStateManager.GetStructMarshallingNativeType((MetadataType)managedType);
+            ThunkType = thunkType;
+        }
+
+        public override TypeSystemContext Context
+        {
+            get
+            {
+                return ManagedType.Context;
+            }
+        }
+
+        public override TypeDesc OwningType
+        {
+            get
+            {
+                return _owningType;
+            }
+        }
+
+        private MethodSignature _signature;
+        public override MethodSignature Signature
+        {
+            get
+            {
+                if (_signature == null)
+                {
+                    TypeDesc[] parameters;
+                    if (ThunkType == StructMarshallingThunkType.Cleanup)
+                    {
+                        parameters = new TypeDesc[1];
+                        parameters[0] = NativeType.MakeByRefType();
+                    }
+                    else
+                    {
+                        parameters = new TypeDesc[2];
```

Comment:
Nit: C# has a better syntax for this: `new TypeDesc[] { ManagedType.MakeByRefType(), NativeType.MakeByRefType() }`